components/consumers/pdf/main.go

```diff
@@ -52,98 +61,64 @@ func main() {
 		}
 		responses = r
 	}
-	result, err := buildPdf(responses)
-	if err != nil {
-		log.Fatal(err)
-	}
 
-	if err = sendToS3(result, bucket, region); err != nil {
-		log.Fatal(err)
+	cleanupRun := func(msg string, cleanup func() error) {
+		if err := cleanup(); err != nil {
+			slog.Error(msg, "error", err)
+		}
 	}
-}
 
-func sendToS3(filename, bucket, region string) error {
-	sess, err := session.NewSession(&aws.Config{Region: aws.String(region)})
+	pw, err := playwright.NewClient()
 	if err != nil {
-		return fmt.Errorf("unable to start session with AWS API: %w", err)
+		log.Fatalf("could not launch playwright: %s", err)
 	}
 
-	// filename is statically defined above
-	//#nosec:G304
-	data, err := os.ReadFile(filename) //#nosec:G304
+	defer cleanupRun("could not stop Playwright: %w", pw.Stop)
+
+	client, err := s3client.NewClient(region)
 	if err != nil {
-		return fmt.Errorf("could not open file: %w", err)
+		log.Fatal(err)
 	}
 
-	uploader := s3manager.NewUploader(sess)
-	_, err = uploader.Upload(&s3manager.UploadInput{
-		Bucket: aws.String(bucket),
-		Key:    aws.String(filename),
-		Body:   bytes.NewReader(data),
-	})
+	if err := run(responses, pw, client); err != nil {
+		log.Fatal(err)
+	}
+}
+
+func run(responses any, pw playwright.Wrapper, s3Wrapper s3client.Wrapper) error {
+	slog.Info("reading pdf")
+	resultFilename, pdfBytes, err := buildPdf(responses, pw)
 	if err != nil {
-		return fmt.Errorf("unable to upload %q to %q: %w", filename, bucket, err)
+		return err
 	}
 
-	fmt.Printf("Successfully uploaded %q to %q\n", filename, bucket)
-	return nil
+	slog.Info("uploading pdf to s3", slog.String("filename", resultFilename))
+	return s3Wrapper.UpsertFile(resultFilename, bucket, pdfBytes)
 }
 
-func buildPdf(data any) (string, error) {
+func buildPdf(data any, pw playwright.Wrapper) (string, []byte, error) {
 	tmpl := template.Must(template.ParseFiles("default.html"))
```

Comment:
⛏️ we could use a `template` function that doesn't panic on error so we can handle the error scenario better